src/utility/svd.jl

```diff
@@ -8,53 +8,106 @@ using TensorKit: SectorDict
+        truncerr = zero(real(scalartype(t)))
+        return _empty_svdtensors(t)..., truncerr
+    end
+
+    Udata, Σdata, Vdata, dims = _compute_svddata!(t, alg, trunc)
+    U, S, V = _create_svdtensors(t, Udata, Σdata, Vdata, spacetype(t)(dims))
+    truncerr = trunc isa NoTruncation ? abs(zero(scalartype(t))) : norm(U * S * V - t, p)
 
-# function TensorKit._compute_svddata!(t::TensorMap, alg::IterSVD)
-#     InnerProductStyle(t) === EuclideanProduct() || throw_invalid_innerproduct(:tsvd!)
-#     I = sectortype(t)
-#     A = storagetype(t)
-#     Udata = SectorDict{I,A}()
-#     Vdata = SectorDict{I,A}()
-#     dims = SectorDict{I,Int}()
-#     local Σdata
-#     for (c, b) in blocks(t)
-#         x₀ = randn(eltype(b), size(b, 1))
-#         Σ, lvecs, rvecs, info = KrylovKit.svdsolve(b, x₀, alg.howmany, :LR, alg.alg)
-#         if info.converged < alg.howmany  # Fall back to dense SVD if not properly converged
-#             U, Σ, V = TensorKit.MatrixAlgebra.svd!(b, TensorKit.SVD())
-#             Udata[c] = U
-#             Vdata[c] = V
-#         else
-#             Udata[c] = stack(lvecs)
-#             Vdata[c] = stack(rvecs)'
-#         end
-#         if @isdefined Σdata # cannot easily infer the type of Σ, so use this construction
-#             Σdata[c] = Σ
-#         else
-#             Σdata = SectorDict(c => Σ)
-#         end
-#         dims[c] = length(Σ)
-#     end
-#     return Udata, Σdata, Vdata, dims
-# end
+    return U, S, V, truncerr
+end
+function TensorKit._compute_svddata!(
+    t::TensorMap,
+    alg::IterSVD,
+    trunc::Union{TensorKit.NoTruncation,TensorKit.TruncationSpace},
+)
+    InnerProductStyle(t) === EuclideanProduct() || throw_invalid_innerproduct(:tsvd!)
+    I = sectortype(t)
+    A = storagetype(t)
+    Udata = SectorDict{I,A}()
+    Vdata = SectorDict{I,A}()
+    dims = SectorDict{I,Int}()
+    local Σdata
+    for (c, b) in blocks(t)
+        x₀ = randn(eltype(b), size(b, 1))
+        howmany = trunc isa NoTruncation ? minimum(size(b)) : blockdim(trunc.space, c)
+        Σ, lvecs, rvecs, info = KrylovKit.svdsolve(b, x₀, howmany, :LR, alg.alg)
+        if info.converged < howmany  # Fall back to dense SVD if not properly converged
+            U, Σ, V = TensorKit.MatrixAlgebra.svd!(b, TensorKit.SVD())
+            Udata[c] = U
+            Vdata[c] = V
+        else
+            Udata[c] = stack(lvecs)
+            Vdata[c] = stack(rvecs)'
+        end
+        if @isdefined Σdata # cannot easily infer the type of Σ, so use this construction
+            Σdata[c] = Σ
+        else
+            Σdata = SectorDict(c => Σ)
+        end
+        dims[c] = length(Σ)
+    end
+    return Udata, Σdata, Vdata, dims
+end
 
+# IterSVD adjoint for tsvd! using KrylovKit.svdsolve adjoint machinery for each block
 function ChainRulesCore.rrule(
-    ::typeof(TensorKit.tsvd),
+    ::typeof(TensorKit.tsvd!),
     t::AbstractTensorMap;
     trunc::TruncationScheme=notrunc(),
     p::Real=2,
     alg::IterSVD=IterSVD(),
```

Comment:
you probably should not have the default arguments in the chainrules, they should be populated in the forwards pass, then dispatch to the full method, for which the chainrules are defined

Reply:
Yes good point, but I think this will not be possible using `TensorKit.tsvd` since it doesn't dispatch on the algorithm type (only `_tsvd!` does). I thought I could get away with not using any SVD wrapper but maybe I need to define one after all to be able to dispatch the adjoint on the `alg` type.

Reply:
Ah I was already fearing this. In principle it would still work, we'd have to move the chainrules definitions in TensorKit from `tsvd!` to `_tsvd!`, and then we can still intercept... Maybe the most elegant for now is to just define `PEPSKit.tsvd!` to do what we want from `TensorKit.tsvd!`, and then migrate that upstream once we're happy with the interface?

Reply:
That's precisely what I did now, I guess that works for the moment!